### DISCUSSION TITLE
2024/penzu-export/

### DISCUSSION THREAD

**Discussion by giscus** [mirontoli/chuvash-eu-comments#5] [General]
# 2024/penzu-export/

I have used Penzu as my main journal app for many years. Recently when Apple launched its Journal app I have been looking at it and other competitors. Then I realized that I am not able to get my own data from Penzu. There is no reasonable Export function.
So I found my own way to get my journal data. I could name this blog post to something like “How to export the unexportable”, or “How to intercept XHR requests in Puppeteer”, but my case is about Penzu, so I’ll stick to this particular title.

https://chuvash.eu/2024/penzu-export/

**Comment by jasonblewis** [comment]
This is great! thanks for sharing. I used it to export my penzu notes. I got chat GPT to write me a little python script to convert the notes from richtext to markdown.

``` python
import sys
import site

site.addsitedir('/home/user/.local/pipx/venvs/markdownify/lib/python3.11/site-packages')

from markdownify import markdownify as md
import json
from datetime import datetime

def read_and_process_json(input_file, output_file):
    # Read the JSON file
    with open(input_file, 'r') as f:
        data = json.load(f)

    # Create a dictionary to store entries by id, preferring those with richtext_body
    entries_by_id = {}
    for item in data:
        item_id = item.get("id", "")
        if item_id in entries_by_id:
            # If current item has richtext_body, replace the existing entry
            if item.get("richtext_body", "").strip():
                entries_by_id[item_id] = item
        else:
            entries_by_id[item_id] = item

    # Get the deduplicated entries
    deduplicated_data = list(entries_by_id.values())

    # Sort the data in reverse date order
    deduplicated_data.sort(key=lambda x: datetime.strptime(x['created_at'], '%Y-%m-%dT%H:%M:%SZ'), reverse=True)

    # Write the processed data to the output file
    with open(output_file, 'w') as f:
        for item in deduplicated_data:
            date_str = datetime.strptime(item.get("created_at", ""), '%Y-%m-%dT%H:%M:%SZ').strftime('%Y-%m-%d %H:%M:%S')
            title = item.get("title", "")
            richtext_body = item.get("richtext_body", "")
            plaintext_body = item.get("plaintext", "")
            item_id = item.get("id", "")
            
            # Use richtext_body if available and contains non-whitespace characters, otherwise use plaintext_body
            if richtext_body and richtext_body.strip():
                markdown_body = md(richtext_body).strip()
            else:
                markdown_body = plaintext_body.strip()
            
            if title:
                f.write(f"**{date_str} {title}**\n")
            else:
                f.write(f"**{date_str}**\n")
            
            f.write(f"Penzu ID: {item_id}\n")
            f.write(f"{markdown_body}\n")
            f.write("\n")

if __name__ == "__main__":
    input_file = 'posts.json'
    output_file = 'output.md'
    read_and_process_json(input_file, output_file)
```

**Reply by mirontoli** [reply]
Thank you for your feedback. I am glad it helped you. I also appreciate your markdown conversion script. In my case I only wanted the plain text. But often bold, italics, links are important in a journal, so converting the journal to markdown is the best way to preserve it.

**Comment by erens-basement** [comment]
Did you end up deciding on a new journal app to move to? I'm still considering whether to move to the Apple Journal app but I wish it had an easier way of exporting to HTML or MD.

**Reply by mirontoli** [reply]
Hi! I ended up journaling in the Apple Journal app. It's convinient (when you can embed your pictures, workouts and so on). I hope there will be an option to export the journals in the future. My exported penzu journal is archived on my NAS.

**Reply by erens-basement** [reply]
Thanks for the fast response. I looked into export options, and apparently through Settings > Apps > Journal you can export all entries to HTML which is incredible. I'm sure someone down the line will build a tool to parse it to MD, txt, etc. 

I noticed there's only an iphone app for now, do you exclusively journal through your phone or have another workflow?

**Comment by awysocki** [comment]
THANK YOU for this,  I was able to get it working. They now use oauth so I had to make some changes but I also wrote a routine to convert the download items into Evernote format .ENEX  
My project is https://github.com/awysocki/Penzu_2_ENEX  

So you can end up with 1 file that many journaling sites accept.

Again THANK YOU for all you did, it helped so much.
